More robust logic for xlsx sheet lookup; fixes #104, relates to #80

jennybc opened the pull request:
Current lookup assumes that xml for sheet i is in `xl/worksheets/sheeti.xml`. But it's not that simple. Instead we need to consult the i-th `sheet` node in `xl/workbook.xml` to learn the sheet's `Id`, e.g. `rId5`. Then lookup the target xml file based on this `Id` in `xl/_rels/workbook.xml.rels`.

When does this matter? Various non-Excel tools write xlsx affected by this. And it even happens in xlsx written by Excel if there are embedded chartsheets.

Thx @jimhester for writing most of this while we worked together last week.

Real example of `i` vs worksheet name vs `Id` vs xml target file (reflects test workbook sheet-xml-lookup.xlsx):

``` r
## i      name    Id                   Target
##      <chr> <chr>                    <chr>
## 1   Africa  rId3 xl/worksheets/sheet4.xml
## 2 Americas  rId4 xl/worksheets/sheet3.xml
## 3     Asia  rId5 xl/worksheets/sheet5.xml
## 4   Europe  rId6 xl/worksheets/sheet1.xml
## 5  Oceania  rId7 xl/worksheets/sheet2.xml
```

Browse the xml for this sheet [mini-gap.xlsx](https://jennybc.github.io/excelgesis/mini-gap/index.html).

hadley commented on src/XlsxWorkBook.h:
```diff
@@ -9,56 +9,126 @@
 
 class XlsxWorkBook {
+
+  class SheetRelations {
```
Might be useful to add a brief comment here, summarising the purpose of this class

jennybc replied:
done in 1fa3cae

hadley commented on src/XlsxWorkBook.h:
```diff
@@ -9,56 +9,126 @@
+      }
+    }
+
+    void parse_workbook_rels(const std::string path) {
```
Missing `&`

jennybc replied:
done in f110164

hadley commented on tests/testthat/test-sheets.R:
```diff
@@ -9,3 +9,47 @@ test_that("excel_sheets returns utf-8 encoded text", {
 })
+
+test_that("informative error when requesting non-existent sheet by name", {
+  expect_error(read_excel("iris-excel.xlsx", sheet = "tulip"),
```
Can you please indent long lines like:

```R
foo(
  a,
  b
)
```

jennybc replied:
done in 55ded75

hadley commented on tests/testthat/test-sheets.R:
```diff
@@ -9,3 +9,47 @@ test_that("excel_sheets returns utf-8 encoded text", {
+
+## #104, #168, some of #80
+## #116, #200 (chartsheet case)
+test_that("sheet data xml target is explicitly looked up [xlsx]", {
```
We have a mild convention to put issue numbers in the test name: "looked up (#104, #168)" ...

You might also consider consolidating all those issues into one so you can just refer to a single number.

jennybc replied:
I have made #104 the main issue for this now.

jennybc replied:
done in 55ded75

jennybc commented:
I removed `tests/testthat/sheet-xml-lookup.xls` (97c12a5) and addressed the other comments.

I'm going to interpret "otherwise, looks good" as permission to merge this myself.
